### PR TITLE
Gemma4 tool calling grammar

### PR DESCRIPTION
**Pull request by neilmehta24**
## Summary

This PR adds a Gemma4-only tool-call logits processor for the MLX Engine batched VLM generator.

It activates only when a Gemma4 VLM prompt already contains native Gemma4 tool declarations. If the model later chooses to emit `<|tool_call>`, MLX Engine constrains the rest of the native call so malformed structure is not sampled into the generated token stream.

## Why

MLX Engine currently streams model-native text and applies ordinary stop/string handling after tokens are sampled. That is not enough for Gemma4 tool-call reliability: once a model starts a native call, invalid syntax can create an invalid tool call, which hurts model effectiveness.

The main reliability target here is structural correctness after the model has already selected tool-call mode:

```text
<|tool_call>call:get_weather{location:<|"|>Paris<|"|>}<tool_call|>
```

The processor prevents common bad continuations such as unknown tool names, broken `call:` headers, unbalanced argument objects, unclosed Gemma strings, missing `<tool_call|>`.

## Design decisions and tradeoffs

- **Reasoning guard**: vLLM’s Gemma4 parser can treat `<|tool_call>` inside reasoning as an implicit transition out of reasoning. Since the LM Studio app currently ignores tool calls within reasoning blocks, update the model logits to disallow it. This will materially change the model output generation, but overall prevents models from making tool calls without prior structure. Down the road, when we can handle tool calls interrupting thinking, we can refactor to remove this logit mask.
- **llguidance grammar**: llguidance gives us a compact grammar for the post-`<|tool_call>` span and only runs once tool-call mode has started. The grammar enforces structure and known tool names.
- **Preserve adjacent calls**: After `<tool_call|>`, the processor does not force an immediate stop. It allows whitespace and another adjacent `<|tool_call>`, so multiple back-to-back native calls remain possible if the model emits them.

## Performance / regression considerations

Normal generation should remain effectively unchanged. A small amount of extra work while actively decoding a tool call is acceptable; the important path is no-tool / pre-tool generation. No-tool perf smoke on the Gemma4 E2B VLM path showed no material slowdown:

  ```text
  baseline median: 93.69 tok/s
  guarded median:  93.22 tok/s
  delta: -0.51%
  ```

## References
Upstream behavior used for comparison:

- vLLM Gemma4 parser preserves special tokens, tracks visible reasoning, parses `<|tool_call>call:NAME{...}<tool_call|>`, and scans args until `<tool_call|>` outside Gemma strings: https://github.com/vllm-project/vllm/blob/main/rust/src/parser/src/unified/gemma4.rs
- llama.cpp Gemma4 chat grammar preserves Gemma4 markers, defines Gemma4 thinking tags, triggers grammar on `<|tool_call>`, and constrains native calls as `<|tool_call>call:...<tool_call|>`: https://github.com/ggml-org/llama.cpp/blob/master/common/chat.cpp
- llguidance supports Lark-like CFGs, tokenizer special-token references, numeric token-id references, and efficient token-mask computation: https://github.com/guidance-ai/llguidance/blob/main/docs/syntax.md and https://github.com/guidance-ai/llguidance

### CODE REVIEW

File: mlx_engine/tool_runtime.py [hunk 1]
```diff
@@ -0,0 +1,352 @@
...
+            # this step. Skip it there so llguidance does not see it twice.
+            self._context_token_count += 1
+            if self._tool_state == self._STATE_POST_TOOL:
+                # The call just closed. Only allow EOS, whitespace, or another
+                # adjacent <|tool_call> from here.
+                logits = _boost_token_ids_mx(
+                    logits,
+                    mx.array(True),
+                    self._post_tool_token_ids,
+                )
```
[the remark this answers is not in the document]
Comment (neilmehta24) [reply mlx_engine/tool_runtime.py]: Good catch, fixed.

File: mlx_engine/tool_runtime.py [hunk 2]
```diff
@@ -0,0 +1,362 @@
...
+        self._reasoning_start_second_token_id = reasoning_start_token_ids[1]
+        self._reasoning_end_token_id = reasoning_end_token_ids[0]
+        self._tool_call_start_token_id = tool_call_start_token_id
+        self._tool_grammar = tool_grammar
+        self._initial_tool_token_ids = tuple(tool_grammar.initial_token_ids)
+        self._post_tool_token_ids = (
+            *eos_token_ids,
+            *whitespace_token_ids,
+            tool_call_start_token_id,
+        )
```
[the remark this answers is not in the document]
Comment (neilmehta24) [reply mlx_engine/tool_runtime.py:127]: > When a caller relies on stop_strings for the Gemma turn terminator

Extreme edge case.

> leave tool-call responses stuck emitting whitespace until max_tokens

Very unlikely. I probed the logits after tool calls, and sampling whitespace for gemma heavily downweights repeated whitespace in the subsequent decodes.

**Comment by will-lms** [comment]
# GPT Review
## [P1] Colon-namespaced tools are silently excluded
At `mlx_engine/tool_runtime.py:24`, `_GEMMA4_TOOL_NAME_RE` permits `._$/-` but not `:`, even though Gemma 4 supports namespaced tool IDs such as `mcp:search`. These declarations are not extracted, so the guard is disabled when all tools are namespaced and makes namespaced calls impossible when mixed with recognized tool names.

**Comment by neilmehta24** [comment]
> # GPT Review
> ## [P1] Colon-namespaced tools are silently excluded
> At `mlx_engine/tool_runtime.py:24`, `_GEMMA4_TOOL_NAME_RE` permits `._$/-` but not `:`, even though Gemma 4 supports namespaced tool IDs such as `mcp:search`. These declarations are not extracted, so the guard is disabled when all tools are namespaced and makes namespaced calls impossible when mixed with recognized tool names.

Pushed a fix for this

**Review by will-lms — APPROVED** [review]
